File: go/ashid.go
package ashid

import (
	"errors"
	"fmt"
	"math/big"
	"strings"
	"time"
)

// MaxTimestamp is the largest timestamp (ms since Unix epoch) that fits in
// 9 Crockford Base32 characters: Dec 12, 3084.
const MaxTimestamp int64 = 35184372088831

const (
	randomEncodedLength    = 13
	timestampEncodedLength = 9
	standardBaseLength     = 22
	ashid4BaseLength       = 26

	// MaxIDLen is enough headroom for any Ashid: longest prefix users typically
	// pick (~32 chars) + 9 timestamp + 13 random + 1 delimiter. Use as the
	// capacity hint for stack-allocated Append buffers.
	MaxIDLen = 64
)

// New returns a time-sortable Ashid. With no argument the result is the
// fixed 22-char base form; with a prefix it returns the variable-length form.
//
// Pass at most one prefix; additional arguments are ignored.
func New(prefix ...string) string {
	var buf [MaxIDLen]byte
	return string(AppendNew(buf[:0], prefix...))
}

// New4 returns an Ashid with two random components (UUID v4 equivalent).
// The base ID is always 26 chars (13 + 13 = ~106 bits of entropy).
//
// Pass at most one prefix; additional arguments are ignored.
func New4(prefix ...string) string {
	var buf [MaxIDLen]byte
	return string(AppendNew4(buf[:0], prefix...))
}

// AppendNew appends a freshly-generated time-sortable Ashid to dst and returns
// the extended slice. Avoids the string allocation of New for callers writing
// into a pooled buffer.
//
// Pass at most one prefix; additional arguments are ignored.
func AppendNew(dst []byte, prefix ...string) []byte {
	p := ""
	if len(prefix) > 0 {
		p = prefix[0]
	}
	out, _ := Append(dst, p, nowMs(), SecureRandomLong())
	return out
}

// AppendNew4 appends a freshly-generated UUID-v4-style Ashid to dst.
//
// Pass at most one prefix; additional arguments are ignored.
func AppendNew4(dst []byte, prefix ...string) []byte {
	p := ""
	if len(prefix) > 0 {
		p = prefix[0]
	}
	out, _ := Append4(dst, p, SecureRandomLong(), SecureRandomLong())
	return out
}

// Create builds an Ashid from explicit components. Use New for the common case.
//
// Format: [prefix_][timestamp][random]
//
// Prefix rules:
//   - Non-alphanumeric characters are stripped from the prefix; the result is
//     lowercased and a trailing "_" is appended automatically.
//   - An empty prefix (after stripping) produces the no-prefix fixed form.
//
// Timestamp must be in [0, MaxTimestamp]. Random is a non-negative uint64.
func Create(prefix string, timestampMs int64, randomLong uint64) (string, error) {
	var buf [MaxIDLen]byte
	out, err := Append(buf[:0], prefix, timestampMs, randomLong)
	if err != nil {
		return "", err
	}
	return string(out), nil
}

// Create4 builds a UUID-v4-style Ashid from explicit random components.
// Both components are encoded with 13-char zero padding for a 26-char base ID.
func Create4(prefix string, random1, random2 uint64) (string, error) {
	var buf [MaxIDLen]byte
	out, err := Append4(buf[:0], prefix, random1, random2)
	if err != nil {
		return "", err
	}
	return string(out), nil
}

// Append builds an Ashid from explicit components and appends it to dst,
// returning the extended slice. Equivalent to Create but avoids the string
// allocation when writing into a pooled buffer.
func Append(dst []byte, prefix string, timestampMs int64, randomLong uint64) ([]byte, error) {
	if timestampMs < 0 {
		return dst, errors.New("ashid timestamp must be non-negative")
	}
	if timestampMs > MaxTimestamp {
		return dst, fmt.Errorf("ashid timestamp must not exceed %d (Dec 12, 3084)", MaxTimestamp)
	}

	prefixStart := len(dst)
	dst = appendNormalizedPrefix(dst, prefix)
	hasPrefix := len(dst) > prefixStart

	if hasPrefix {
		if timestampMs > 0 {
			dst = AppendEncode(dst, uint64(timestampMs), false)
			dst = AppendEncode(dst, randomLong, true)
		} else {
			dst = AppendEncode(dst, randomLong, false)
		}
	} else {
		dst = appendEncodeWidth(dst, uint64(timestampMs), true, timestampEncodedLength)
		dst = AppendEncode(dst, randomLong, true)
	}
	return dst, nil
}

// Append4 builds a UUID-v4-style Ashid from explicit components and appends
// it to dst.
func Append4(dst []byte, prefix string, random1, random2 uint64) ([]byte, error) {
	dst = appendNormalizedPrefix(dst, prefix)
	dst = AppendEncode(dst, random1, true)
	dst = AppendEncode(dst, random2, true)
	return dst, nil
}

// Parse splits an Ashid into its three components: prefix (with trailing "_"
// or empty), encoded timestamp, and encoded random. Dashes in the prefix
// delimiter are normalized to underscores.
func Parse(id string) (prefix, encodedTimestamp, encodedRandom string, err error) {
	if id == "" {
		err = errors.New("invalid ashid: cannot be empty")
		return
	}

	prefixLength := 0
	hasDelimiter := false

	for i := 0; i < len(id); i++ {
		ch := id[i]
		if isAlpha(ch) {
			prefixLength++
		} else if (ch == '_' || ch == '-') && prefixLength > 0 {
			prefixLength++
			hasDelimiter = true
			break
		} else {
			prefixLength = 0
			break
		}
	}

	if hasDelimiter {
		raw := id[:prefixLength]
		if strings.HasSuffix(raw, "-") {
			raw = raw[:len(raw)-1] + "_"
		}
		prefix = raw
	}

	baseID := id[prefixLength:]
	if baseID == "" {
		err = errors.New("invalid ashid: must have a base ID")
		return
	}

	switch {
	case hasDelimiter:
		if len(baseID) <= randomEncodedLength {
			encodedTimestamp = "0"
			encodedRandom = baseID
		} else {
			encodedTimestamp = baseID[:len(baseID)-randomEncodedLength]
			encodedRandom = baseID[len(baseID)-randomEncodedLength:]
		}
	case len(baseID) == ashid4BaseLength:
		encodedTimestamp = baseID[:randomEncodedLength]
		encodedRandom = baseID[randomEncodedLength:]
	case len(baseID) == standardBaseLength:
		encodedTimestamp = baseID[:timestampEncodedLength]
		encodedRandom = baseID[timestampEncodedLength:]
	default:
		err = fmt.Errorf("invalid ashid: base ID must be %d or %d characters without delimiter (got %d)",
			standardBaseLength, ashid4BaseLength, len(baseID))
	}
	return
}

// Prefix returns the prefix (with trailing "_") or "" if none.
func Prefix(id string) (string, error) {
	p, _, _, err := Parse(id)
	return p, err
}

// Timestamp returns the embedded timestamp in milliseconds since Unix epoch.
func Timestamp(id string) (int64, error) {
	_, ts, _, err := Parse(id)
	if err != nil {
		return 0, err
	}
	v, err := Decode(ts)
	if err != nil {
		return 0, err
	}
	return int64(v), nil
}

// Random returns the embedded random component as a *big.Int. Using big.Int
// preserves full precision for IDs whose random component encodes more than
// 64 bits (mirrors the BigInt return in TypeScript and arbitrary-precision int
// in Python).
func Random(id string) (*big.Int, error) {
	_, _, r, err := Parse(id)
	if err != nil {
		return nil, err
	}
	return DecodeBigInt(r)
}

// IsValid reports whether s is a well-formed Ashid.
func IsValid(id string) bool {
	if id == "" {
		return false
	}
	prefix, ts, r, err := Parse(id)
	if err != nil {
		return false
	}
	if prefix != "" && !validPrefix(prefix) {
		return false
	}
	if !IsValidBase32(ts) || !IsValidBase32(r) {
		return false
	}
	return true
}

// Normalize lowercases the prefix and re-encodes the ID through canonical form,
// mapping ambiguous characters (I/L -> 1, O -> 0, U -> V) along the way.
func Normalize(id string) (string, error) {
	prefix, ts, r, err := Parse(id)
	if err != nil {
		return "", err
	}
	timestamp, err := Decode(ts)
	if err != nil {
		return "", err
	}
	random, err := DecodeBigInt(r)
	if err != nil {
		return "", err
	}

	var normalizedPrefix string
	if prefix != "" {
		normalizedPrefix = strings.ToLower(prefix)
	}

	if random.IsUint64() {
		return Create(normalizedPrefix, int64(timestamp), random.Uint64())
	}
	return createBigInt(normalizedPrefix, int64(timestamp), random)
}

// createBigInt mirrors Create but accepts a *big.Int random for full-precision
// round-trips. Used by Normalize for IDs whose random component exceeds 64 bits.
func createBigInt(prefix string, timestampMs int64, randomLong *big.Int) (string, error) {
	if timestampMs < 0 {
		return "", errors.New("ashid timestamp must be non-negative")
	}
	if timestampMs > MaxTimestamp {
		return "", fmt.Errorf("ashid timestamp must not exceed %d (Dec 12, 3084)", MaxTimestamp)
	}
	if randomLong == nil || randomLong.Sign() < 0 {
		return "", errors.New("ashid random value must be non-negative")
	}

	var buf [MaxIDLen]byte
	dst := buf[:0]

	prefixStart := len(dst)
	dst = appendNormalizedPrefix(dst, prefix)
	hasPrefix := len(dst) > prefixStart

	if hasPrefix {
		if timestampMs > 0 {
			dst = AppendEncode(dst, uint64(timestampMs), false)
			out, err := AppendEncodeBigInt(dst, randomLong, true)
			if err != nil {
				return "", err
			}
			dst = out
		} else {
			out, err := AppendEncodeBigInt(dst, randomLong, false)
			if err != nil {
				return "", err
			}
			dst = out
		}
	} else {
		dst = appendEncodeWidth(dst, uint64(timestampMs), true, timestampEncodedLength)
		out, err := AppendEncodeBigInt(dst, randomLong, true)
		if err != nil {
			return "", err
		}
		dst = out
	}
	return string(dst), nil
}

func appendNormalizedPrefix(dst []byte, prefix string) []byte {
	if prefix == "" {
		return dst
	}
	start := len(dst)
	for i := 0; i < len(prefix); i++ {
		ch := prefix[i]
		switch {
		case ch >= 'a' && ch <= 'z':
			dst = append(dst, ch)
		case ch >= 'A' && ch <= 'Z':
			dst = append(dst, ch+32)
		case ch >= '0' && ch <= '9':
			dst = append(dst, ch)
		}
	}
	if len(dst) == start {
		return dst
	}
	return append(dst, '_')
}

func validPrefix(prefix string) bool {
	if !strings.HasSuffix(prefix, "_") {
		return false
	}
	body := prefix[:len(prefix)-1]
	if body == "" {
		return false
	}
	for i := 0; i < len(body); i++ {
		ch := body[i]
		ok := (ch >= 'a' && ch <= 'z') ||
			(ch >= 'A' && ch <= 'Z') ||
			(ch >= '0' && ch <= '9')
		if !ok {
			return false
		}
	}
	return true
}

func isAlpha(ch byte) bool {
	return (ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z')
}

func nowMs() int64 {
	return time.Now().UnixMilli()
}

File: python/src/ashid/_ashid.py
"""Ashid - time-sortable unique identifier with optional type prefix.

Format: [prefix_][timestamp][random]

Prefix Rules:
- Prefix must be alphanumeric only (a-z, A-Z, 0-9 — non-alphanumeric chars are stripped)
- Delimiter (_) is automatically added — callers should NOT include it
- If a delimiter is passed in, it is ignored (backward compatibility)
- With prefix: variable length base ID (no timestamp padding when 0)
- Without prefix: fixed 22-char base ID

Timestamp Support:
- Minimum: 0 (Unix epoch, Jan 1, 1970)
- Maximum: 35184372088831 (Dec 12, 3084)
"""

import re
import time as _time
from typing import Optional, Tuple

from ._encoder import EncoderBase32Crockford


_MAX_TIMESTAMP = 35184372088831
_RANDOM_ENCODED_LENGTH = 13
_TIMESTAMP_ENCODED_LENGTH = 9
_STANDARD_BASE_LENGTH = 22
_ASHID4_BASE_LENGTH = 26

_PREFIX_CLEAN_RE = re.compile(r"[^a-zA-Z0-9]")
_PREFIX_VALID_RE = re.compile(r"^[a-zA-Z0-9]+_$")


def _now_ms() -> int:
    return _time.time_ns() // 1_000_000


def _normalize_prefix(prefix: Optional[str]) -> Optional[str]:
    if prefix is None or prefix == "":
        return None
    cleaned = _PREFIX_CLEAN_RE.sub("", prefix).lower()
    if cleaned == "":
        return None
    return cleaned + "_"


def _build_base(prefix: Optional[str], n1: int, n2: int, padded: bool) -> str:
    """Encode two non-negative ints into the canonical base ID form.

    Mirrors the TypeScript buildBase routed through create()/create4():
      - padded=True   -> both halves 13-char zero-padded (ashid4 shape).
      - padded=False  -> first half encoded unpadded when a prefix is present;
                         else padded to 9 chars (timestamp width) when the raw
                         encoding fits in 9 chars, else padded to 13.
    The second half is always 13-char padded.
    """
    if n1 < 0 or n2 < 0:
        raise ValueError("Ashid random value must be non-negative")
    normalized_prefix = _normalize_prefix(prefix)

    if normalized_prefix or padded:
        encoded1 = EncoderBase32Crockford.encode(n1, padded=padded)
    else:
        raw = EncoderBase32Crockford.encode(n1)
        width = (
            _TIMESTAMP_ENCODED_LENGTH
            if len(raw) <= _TIMESTAMP_ENCODED_LENGTH
            else _RANDOM_ENCODED_LENGTH
        )
        encoded1 = raw.rjust(width, "0")
    encoded2 = EncoderBase32Crockford.encode(n2, padded=True)
    return (normalized_prefix or "") + encoded1 + encoded2


class Ashid:
    """Time-sortable unique identifier with optional type prefix."""

    @staticmethod
    def create(
        prefix: Optional[str] = None,
        time: Optional[int] = None,
        random_long: Optional[int] = None,
    ) -> str:
        """Create a new Ashid."""
        normalized_prefix = _normalize_prefix(prefix)

        if time is None:
            time = _now_ms()
        floored_time = int(time)
        if floored_time < 0:
            raise ValueError("Ashid timestamp must be non-negative")
        if floored_time > _MAX_TIMESTAMP:
            raise ValueError(
                f"Ashid timestamp must not exceed {_MAX_TIMESTAMP} (Dec 12, 3084)"
            )

        if random_long is None:
            random_long = EncoderBase32Crockford.secure_random_long()
        if random_long < 0:
            raise ValueError("Ashid random value must be non-negative")

        if normalized_prefix:
            if floored_time > 0:
                random_encoded = EncoderBase32Crockford.encode(random_long, padded=True)
                base_id = EncoderBase32Crockford.encode(floored_time) + random_encoded
            else:
                random_encoded = EncoderBase32Crockford.encode(random_long)
                base_id = random_encoded
        else:
            time_encoded = EncoderBase32Crockford.encode(floored_time).rjust(
                _TIMESTAMP_ENCODED_LENGTH, "0"
            )
            random_encoded = EncoderBase32Crockford.encode(random_long, padded=True)
            base_id = time_encoded + random_encoded

        return (normalized_prefix or "") + base_id

    @staticmethod
    def create4(
        prefix: Optional[str] = None,
        random1: Optional[int] = None,
        random2: Optional[int] = None,
    ) -> str:
        """Create a random Ashid (UUID v4 equivalent) with two padded random components."""
        normalized_prefix = _normalize_prefix(prefix)

        if random1 is None:
            random1 = EncoderBase32Crockford.secure_random_long()
        if random2 is None:
            random2 = EncoderBase32Crockford.secure_random_long()
        if random1 < 0 or random2 < 0:
            raise ValueError("Ashid random values must be non-negative")

        encoded1 = EncoderBase32Crockford.encode(random1, padded=True)
        encoded2 = EncoderBase32Crockford.encode(random2, padded=True)
        base_id = encoded1 + encoded2

        return (normalized_prefix or "") + base_id

    @staticmethod
    def parse(id: str) -> Tuple[str, str, str]:
        """Parse an Ashid into (prefix_with_trailing_, encoded_timestamp, encoded_random)."""
        if not id:
            raise ValueError("Invalid Ashid: cannot be empty")

        prefix_length = 0
        has_delimiter = False

        for ch in id:
            if ch.isascii() and ch.isalpha():
                prefix_length += 1
            elif (ch == "_" or ch == "-") and prefix_length > 0:
                prefix_length += 1
                has_delimiter = True
                break
            else:
                prefix_length = 0
                break

        # If no delimiter was found, the whole string is the base — even if it
        # happens to be all-alpha (e.g. an ashid4 whose random component encodes
        # to letters only).
        if not has_delimiter:
            prefix_length = 0

        if has_delimiter:
            raw_prefix = id[:prefix_length]
            if raw_prefix.endswith("-"):
                raw_prefix = raw_prefix[:-1] + "_"
            prefix = raw_prefix
        else:
            prefix = ""

        base_id = id[prefix_length:]

        if not base_id:
            raise ValueError("Invalid Ashid: must have a base ID")

        if has_delimiter:
            if len(base_id) <= _RANDOM_ENCODED_LENGTH:
                encoded_timestamp = "0"
                encoded_random = base_id
            else:
                encoded_timestamp = base_id[:-_RANDOM_ENCODED_LENGTH]
                encoded_random = base_id[-_RANDOM_ENCODED_LENGTH:]
        elif len(base_id) == _ASHID4_BASE_LENGTH:
            encoded_timestamp = base_id[:_RANDOM_ENCODED_LENGTH]
            encoded_random = base_id[_RANDOM_ENCODED_LENGTH:]
        elif len(base_id) == _STANDARD_BASE_LENGTH:
            encoded_timestamp = base_id[:_TIMESTAMP_ENCODED_LENGTH]
            encoded_random = base_id[_TIMESTAMP_ENCODED_LENGTH:]
        else:
            raise ValueError(
                f"Invalid Ashid: base ID must be {_STANDARD_BASE_LENGTH} or "
                f"{_ASHID4_BASE_LENGTH} characters without delimiter (got {len(base_id)})"
            )

        return prefix, encoded_timestamp, encoded_random

    @staticmethod
    def prefix(id: str) -> str:
        """Extract the prefix (with trailing _) from an Ashid; empty string if none."""
        return Ashid.parse(id)[0]

    @staticmethod
    def timestamp(id: str) -> int:
        """Extract the timestamp (in ms) from an Ashid."""
        _, encoded_timestamp, _ = Ashid.parse(id)
        return EncoderBase32Crockford.decode(encoded_timestamp)

    @staticmethod
    def random(id: str) -> int:
        """Extract the random component from an Ashid as an int (full 64-bit precision)."""
        _, _, encoded_random = Ashid.parse(id)
        return EncoderBase32Crockford.decode(encoded_random)

    @staticmethod
    def is_valid(id: str) -> bool:
        """Validate whether a string is a well-formed Ashid."""
        if not id:
            return False
        try:
            prefix, encoded_timestamp, encoded_random = Ashid.parse(id)
            if prefix and not _PREFIX_VALID_RE.match(prefix):
                return False
            EncoderBase32Crockford.decode(encoded_timestamp)
            EncoderBase32Crockford.decode(encoded_random)
            return True
        except (ValueError, TypeError):
            return False

    @staticmethod
    def normalize(id: str) -> str:
        """Normalize an Ashid: lowercase prefix and re-encode through canonical form.

        Type-1 inputs round-trip to type-1 shape; full-entropy ashid4 inputs
        round-trip to ashid4 shape (the first long naturally fills 13 chars).
        An ashid4 with a small first long collapses to type-1 shape — the two
        longs survive, only the string shape changes.
        """
        prefix, encoded_first, encoded_second = Ashid.parse(id)
        normalized_prefix = prefix.lower() if prefix else None
        long1 = EncoderBase32Crockford.decode(encoded_first)
        long2 = EncoderBase32Crockford.decode(encoded_second)
        return _build_base(normalized_prefix, long1, long2, False)

    @staticmethod
    def to_uuid(id: str) -> str:
        """Convert an Ashid to its UUID-shaped representation.

        An Ashid encodes 128 bits of information (a 64-bit timestamp or first random
        component plus a 64-bit random component); this method emits those 128 bits
        as a standard 36-character UUID string with dashes (8-4-4-4-12).

        Round-trips losslessly with Ashid.from_uuid. The resulting UUID is shape-
        compatible with RFC 4122 but the version/variant bits reflect the underlying
        Ashid bytes, not RFC 4122 conventions, unless the Ashid was originally
        created from a v1/v4/v7 UUID.
        """
        _, encoded_timestamp, encoded_random = Ashid.parse(id)
        high = EncoderBase32Crockford.decode(encoded_timestamp)
        low = EncoderBase32Crockford.decode(encoded_random)
        high_hex = f"{high:016x}"
        low_hex = f"{low:016x}"
        return f"{high_hex[:8]}-{high_hex[8:12]}-{high_hex[12:16]}-{low_hex[:4]}-{low_hex[4:16]}"

    @staticmethod
    def from_uuid(uuid: str, prefix: Optional[str] = None) -> str:
        """Convert a UUID into an Ashid.

        Splits the 128-bit UUID into two 64-bit halves: the high half becomes the
        timestamp (or first random component, see below) and the low half becomes
        the random component.

        - If the high half fits in 45 bits (<= MAX_TIMESTAMP), the result is a
          standard 22-char Ashid base.
        - Otherwise -- for UUIDv4 (random) and UUIDv7 (whose version bits force
          the high half above 2^45) -- the result is a 26-char ashid4 base.

        Round-trip through Ashid.to_uuid is byte-identical in both cases.
        """
        hex_str = uuid.replace("-", "").lower()
        if len(hex_str) != 32 or any(c not in "0123456789abcdef" for c in hex_str):
            raise ValueError(f'Invalid UUID: must be 32 or 36 hex characters (got "{uuid}")')
        high = int(hex_str[:16], 16)
        low = int(hex_str[16:32], 16)
        if high <= _MAX_TIMESTAMP:
            return Ashid.create(prefix, high, low)
        return Ashid.create4(prefix, high, low)


def ashid(prefix: Optional[str] = None) -> str:
    """Create a new Ashid with optional type prefix."""
    return Ashid.create(prefix)


def ashid4(prefix: Optional[str] = None) -> str:
    """Create a UUID-v4-style Ashid with two padded random components."""
    return Ashid.create4(prefix)


def parse_ashid(id: str) -> Tuple[str, str, str]:
    """Parse an Ashid into (prefix, encoded_timestamp, encoded_random)."""
    return Ashid.parse(id)
